Introduce and use Theme singleton

jarolrod opened the pull request:
Replaces #109 

This introduces a `Theme` singleton which contains color definitions so that colors can be used by name instead of by hex. Additionally, it adjusts color definitions depending on if the gui is in dark or light mode. 

Color definitions are based on our color palettes as defined in our [design system](https://www.figma.com/file/GaCoOSNHB2yMB9ThiDtred/Bitcoin-Core-App-Main?node-id=2643%3A83514): 
![color-definitions](https://user-images.githubusercontent.com/23396902/148453667-bdaa347f-b4e9-4db9-9a82-027a10dc9e05.png)

A convenient demo of the changes can be found here: [theme singleton demo](https://github.com/jarolrod/gui-qml/tree/theme-singleton-demo)

This demo adds a `ThemeToggleButton` so that we can toggle between light and dark. Uses the `moon` and `sun` icon from [BitcoinIcons](https://github.com/BitcoinDesign/Bitcoin-Icons).

shaavan commented on src/qml/controls/Theme.qml:
```diff
@@ -0,0 +1,27 @@
+        property color green: dark ? "#36B46B" : "#27AE60"
+        property color blue: dark ? "#3CA3DE" : "#2D9CDB"
+        property color purple: dark ? "#C075DC" : "#BB6BD9"
+        property color neutral0: dark ? "#FFFFFF" : "#000000"
+        property color neutral1: dark ? "#F8F8F8" : "#1A1A1A"
+        property color neutral2: dark ? "#F4F4F4" : "#2D2D2D"
+        property color neutral3: dark ? "#EDEDED" : "#444444"
+        property color neutral4: dark ? "#DEDEDE" : "#5C5C5C"
+        property color neutral5: dark ? "#BBBBBB" : "#787878"
+        property color neutral6: dark ? "#999999" : "#949494"
+        property color neutral7: dark ? "#777777" : "#B0B0B0"
+        property color neutral8: dark ? "#404040" : "#CCCCCC"
+        property color neutral9: dark ? "#000000" : "#FFFFFF"
```
I want to point out one thing. I think you have swapped the color values as they are present in the design system for neutral(0-9) colors.

**Example:**
In design system:
![Screenshot from 2022-04-06 13-52-06](https://user-images.githubusercontent.com/85434418/161930434-f4cdfb81-ffb9-46c9-8df1-52b631d730a9.png)
neutral1 should be `#1A1A1A` when dark mode is on. Whereas according to the code, it would be `#F8F8F8`.

GBKS replied:
Confirming that they are flipped.

jarolrod replied:
Thanks for the eagle eye review, addressed

shaavan reviewed (commented):
Concept ACK

- The `Theme.qml` file and `qmldir` file are introduced correctly.

I was able to find some corrections that will help improve this PR.

Places where Theme singleton could be used:

- In ContinueButton.qml file. `color: "white"`
- In OptionSwitch.qml file. `color: "#ffffff"`

If I am mistaken, and the above colors should remain unchanged with dark mode toggling, I would suggest adding a non-changing white color variant in Theme.qml file as well and using it.

GBKS commented:
Agree with @shaavan's suggestion. There are just a couple of places where I think we want to have a consistent color across both themes, the inactive toggle state being one of them.

I updated the Figma file with a "Static white" color and applied to the toggle. Let me know if you see other places to update this, and other colors to add.

jarolrod commented:
Updated from ([pr114.01](https://github.com/jarolrod/gui-qml/tree/pr114.01) -> [pr114.02](https://github.com/jarolrod/gui-qml/tree/pr114.02), [compare](https://github.com/jarolrod/gui-qml/compare/pr114.01..pr114.02)) to address review feedback. Then, updated from [7698dca ](https://github.com/bitcoin-core/gui-qml/commit/7698dca44d84dbc99e3b2501a875323c30cf1f34)to [581961a](https://github.com/bitcoin-core/gui-qml/commit/581961a82e0edc430e3654697137b1f7db882e34) ([pr114.02](https://github.com/jarolrod/gui-qml/tree/pr114.02) -> [pr114.03](https://github.com/jarolrod/gui-qml/tree/pr114.03), [compare](https://github.com/jarolrod/gui-qml/compare/pr114.02..pr114.03)) to fix trailing whitespace, or so I thought I had fixed it. Finally, updated from [581961a ](https://github.com/bitcoin-core/gui-qml/commit/581961a82e0edc430e3654697137b1f7db882e34)to [44a7696 ](https://github.com/bitcoin-core/gui-qml/commit/44a7696b33f7e61b2017caa00fc92d17ba7e6140) ([pr114.03](https://github.com/jarolrod/gui-qml/tree/pr114.03) -> [pr114.04](https://github.com/jarolrod/gui-qml/tree/pr114.04), [compare](https://github.com/jarolrod/gui-qml/compare/pr114.03..pr114.04)) to finally fix the linter.

The main [diff](https://github.com/jarolrod/gui-qml/compare/pr114.01..pr114.02) to review is between [pr114.01](https://github.com/jarolrod/gui-qml/tree/pr114.01) and [pr114.02](https://github.com/jarolrod/gui-qml/tree/pr114.02)

Changes: incorporate review feedback from @shaavan and @GBKS 
  - Address two instances where components were not using the theme definitions
  - The color definitions were in fact flipped. Addressed the flipped definitions and adjusted the colors called by controls/components to fit inline with the specifications of the [main design file]().
  - Introduced a `white` color definition to work as a static white color. This is to be used in places where the color white should not switch as it does with the definition of the color `neutral0`.

Additionally, the demo has been updated in order to display all components that we have which in turn incorporate all of the controls we have. This allows for convenient testing that the color definitions are as they should be. The demo can be found here: [theme singleton demo](https://github.com/jarolrod/gui-qml/tree/theme-singleton-demo)

shaavan commented on src/qml/controls/Theme.qml:
```diff
@@ -0,0 +1,28 @@
+        property color orange: dark ? "#F89B2A" : "#F7931A"
+        property color red: dark ? "#EC6363" : "#EB5757"
+        property color green: dark ? "#36B46B" : "#27AE60"
+        property color blue: dark ? "#2D9CDB" : "#3CA3DE"
+        property color purple: dark ? "#BB6BD9" : "#C075DC"
```
I think, by mistake, you flipped the _dark toggle_ values for blue and purple as well, which were already correct.

**According to design scheme:**
![Screenshot from 2022-04-12 16-10-09](https://user-images.githubusercontent.com/85434418/162942592-5d8694aa-6706-4205-85a8-4cf44f085291.png)

**According to code:**
blue (when dark): `#2D9CDB`
purple (when dark): `#BB6BD9`

jarolrod replied:
thanks, addressed in https://github.com/bitcoin-core/gui-qml/commit/9b9d198407118475d6ba6dd315991f2a42c813b0

shaavan reviewed (commented):
Changes since my last review:

- Ordering of `neutral0` - `neutral9` is fixed.
- Added a static `white` color to the theme file and used it in the qml directory.

The arrangement of `neutral0` - `neutral9` is toggled and is now in line with the design system file.

I was able to verify that this PR is correctly formatted according to the clang format.

There is, however, one thing that needs to be addressed.

jarolrod commented:
updated from [44a7696 ](https://github.com/bitcoin-core/gui-qml/commit/44a7696b33f7e61b2017caa00fc92d17ba7e6140)to [9b9d198](https://github.com/bitcoin-core/gui-qml/commit/9b9d198407118475d6ba6dd315991f2a42c813b0) ([pr114.04](https://github.com/jarolrod/gui-qml/tree/pr114.04) -> [pr114.05](https://github.com/jarolrod/gui-qml/tree/pr114.05), [compare](https://github.com/jarolrod/gui-qml/compare/pr114.04..pr114.05))

Changes: addressed @shaavan comments. Thank you for the eagle-eye review

shaavan reviewed (approved):
ACK 9b9d198407118475d6ba6dd315991f2a42c813b0

Changes since my last review:

- The color values for the dark and light themes for blue and purple have been fixed.

I successfully ran the [theme singleton demo](https://github.com/jarolrod/gui-qml/tree/theme-singleton-demo). I am attaching a few screenshots to show the correct working of this PR.

**Screenshots:**

| |1|2|
|-----|------|-----|
|**Light Theme**|![Light1](https://user-images.githubusercontent.com/85434418/168283442-690b07b6-837c-433f-a5b5-639acd512acf.png)|![Light2](https://user-images.githubusercontent.com/85434418/168283590-ed1183a3-23d3-4c8c-8046-f813522e8c90.png)|
|**Dark Theme**|![Dark1](https://user-images.githubusercontent.com/85434418/168283487-f4eea5d8-a274-4e7f-aad7-31308028fed1.png)|![Dark2](https://user-images.githubusercontent.com/85434418/168283515-c43a8859-1a77-4a3a-9cc0-09b82fb6c643.png)|
